### ISSUE TITLE
End-to-end tracing: No consistent trace in Zipkin or Jaeger: Service invocation in .NET via HTTP -> Pub in .NET via SDK -> Sub in Go via HTTP -> Pub in Go via SDK -> Sub in .NET via SDK

### ISSUE THREAD

**Issue by ChristianWeyer** [dapr/dapr#4571]
Hi all,

we are not sure whether this is a bug - or a given fact (and if the latter, we might be missing the 'automagic end-to-end tracing' promise by Dapr).

@yaron2 - as briefly talked about in the Discord channel.
Here goes...

## In what area(s)?
/area runtime

## What version of Dapr?
edge (1.7.2)

## Expected Behavior

This is the application scenario we are talking about:

![Dapr_sample](https://user-images.githubusercontent.com/888718/165946589-e3012deb-a389-4b8e-9b59-62d52f809a1e.jpg)

We are using service invocation, and pub/sub. We are not using any bindings for now.

So, there is:
- Service invocation in .NET via HTTP ->
- Pub in .NET via SDK ->
- Sub in Go via HTTP ->
- Pub in Go via SDK ->
- Sub in .NET via SDK

We do not see a consistent end-to-end trace in Zipkin or Jager.
Here are the screenshots from Zipkin:

<img width="1270" alt="Screenshot 2022-04-29 at 14 26 47" src="https://user-images.githubusercontent.com/888718/165947241-9dab050e-1fc3-416e-9e29-842a7e8d4fa4.png">

<img width="2537" alt="Screenshot 2022-04-29 at 14 27 02" src="https://user-images.githubusercontent.com/888718/165947262-5cf79106-a614-4558-bd2f-e0296f80ee65.png">

<img width="2539" alt="Screenshot 2022-04-29 at 14 27 21" src="https://user-images.githubusercontent.com/888718/165948421-92e2815c-8510-46ed-86c4-49fa6d66ea9a.png">

As you can see, there is no single end-to-end trace found.

In addition, here are the two JSON files from Zipkin:

```
[
    {
        "traceId": "4569601be7a75eaa9249727dccf2aa83",
        "parentId": "437029e2d578de85",
        "id": "2429902098158c71",
        "kind": "CLIENT",
        "name": "/dapr.proto.runtime.v1.dapr/publishevent",
        "timestamp": 1651235174395313,
        "duration": 7000,
        "localEndpoint": {
            "serviceName": "orders",
            "ipv4": "10.5.0.10"
        },
        "tags": {
            "dapr.api": "/dapr.proto.runtime.v1.Dapr/PublishEvent",
            "dapr.protocol": "grpc",
            "messaging.destination": "new_orders",
            "messaging.destination_kind": "topic",
            "messaging.system": "pubsub",
            "rpc.service": "Dapr"
        }
    },
    {
        "traceId": "4569601be7a75eaa9249727dccf2aa83",
        "parentId": "8f039a3e0cb775fb",
        "id": "9df217aa9cb0fb61",
        "kind": "SERVER",
        "name": "calllocal/orders/orders",
        "timestamp": 1651235174344721,
        "duration": 76274,
        "localEndpoint": {
            "serviceName": "orders",
            "ipv4": "10.5.0.10"
        },
        "tags": {
            "dapr.api": "/dapr.proto.internals.v1.ServiceInvocation/CallLocal",
            "dapr.invoke_method": "orders",
            "dapr.protocol": "grpc",
            "rpc.service": "ServiceInvocation"
        }
    },
    {
        "traceId": "4569601be7a75eaa9249727dccf2aa83",
        "parentId": "090d39eddeac617c",
        "id": "8f039a3e0cb775fb",
        "kind": "CLIENT",
        "name": "calllocal/orders/orders",
        "timestamp": 1651235174331925,
        "duration": 90028,
        "localEndpoint": {
            "serviceName": "gateway",
            "ipv4": "10.5.0.7"
        },
        "tags": {
            "dapr.api": "POST /v1.0/invoke/orders/method/orders",
            "dapr.protocol": "http",
            "dapr.status_code": "202",
            "net.peer.name": "orders",
            "opencensus.status_description": "OK",
            "rpc.service": "ServiceInvocation"
        }
    },
    {
        "traceId": "4569601be7a75eaa9249727dccf2aa83",
        "parentId": "2429902098158c71",
        "id": "e1f62f72424aab81",
        "kind": "SERVER",
        "name": "pubsub/new_orders",
        "timestamp": 1651235174412009,
        "duration": 2039685,
        "localEndpoint": {
            "serviceName": "shipping",
            "ipv4": "10.5.0.11"
        },
        "tags": {
            "messaging.destination": "new_orders",
            "messaging.destination_kind": "topic",
            "messaging.system": "pubsub",
            "opencensus.status_description": "OK"
        }
    }
]
```

```
[
    {
        "traceId": "527ac55b2279682598d3c5465b5a6879",
        "id": "ec0b05f99306a9e6",
        "kind": "CLIENT",
        "name": "/dapr.proto.runtime.v1.dapr/publishevent",
        "timestamp": 1651235176435384,
        "duration": 832,
        "localEndpoint": {
            "serviceName": "shipping",
            "ipv4": "10.5.0.11"
        },
        "tags": {
            "dapr.api": "/dapr.proto.runtime.v1.Dapr/PublishEvent",
            "dapr.protocol": "grpc",
            "messaging.destination": "processed_orders",
            "messaging.destination_kind": "topic",
            "messaging.system": "pubsub",
            "rpc.service": "Dapr"
        }
    },
    {
        "traceId": "527ac55b2279682598d3c5465b5a6879",
        "parentId": "ec0b05f99306a9e6",
        "id": "6696a1e471f9b402",
        "kind": "SERVER",
        "name": "pubsub/processed_orders",
        "timestamp": 1651235176445602,
        "duration": 61340,
        "localEndpoint": {
            "serviceName": "notification",
            "ipv4": "10.5.0.12"
        },
        "tags": {
            "messaging.destination": "processed_orders",
            "messaging.destination_kind": "topic",
            "messaging.system": "pubsub",
            "opencensus.status_description": "OK"
        }
    }
]
```

As you can see, there is no common parent trace id - which would be expected for real end-to-end tracing.

## Actual Behavior
We should see a consistent end-to-end trace.

## Steps to Reproduce the Problem
A sample of the above architecture diagram can be found here.
https://github.com/thinktecture-labs/cloud-native-sample

It should be runnable with `docker compose` locally.

Any insights and help would be appreciated 😊.

Thank you!

Cheers,
-Christian

**Comment by yaron2** [comment]
Dapr generates the trace headers and propagates them for service to service or pub/sub to service calls, but for the scenario you're describing you'd need to extract the trace headers from the incoming call to the app (either pub/sub or service invocation) and put the trace header on the Cloud Event or the call you're making to Dapr with service invocation.

Read here: https://docs.dapr.io/developing-applications/building-blocks/observability/w3c-tracing/w3c-tracing-overview/#you-need-to-propagate-or-generate-trace-context-between-services. (The paragraph above this one in the docs needs to be rephrased).

And here: https://docs.dapr.io/developing-applications/building-blocks/observability/w3c-tracing/w3c-tracing-howto/

**Comment by msfussell** [comment]
This is an issue that needs to be addressed. The difficulty here is that the SDKs could be improved to have overloads that allow for the propagation of the trace headers. For example in the diagram above when the request arrives from the Gateway service, right now you will have to create a new cloudevent and copy the trace headers into this cloud event, rather than letting Dapr create a cloud event on your behalf, where all you want is to be able to insert the trace header that can from the Gateway service. 

In a nutshell each of the SDKs for the APIs (pud/sub, etc) there needs to be an overload that takes a trace header to allow this to be propagated more easily. [and Dapr cannot do this propagation for you since Dapr  cannot see the application code and the relationships between the events, so this is a developer task.]

**Comment by yaron2** [comment]
The Gateway Service calls the Order Service. This will have Dapr generate the trace ID for the call. In the Orders Service HTTP endpoint, you need to inspect the request and get the Trace ID header. To publish a message using that trace Id, you need to use the SDK method to publish your own Cloud Event which contains the `traceid` field. This will then show up on the Cloud Event on subscribers, and need to be passed in via custom Cloud Event envelopes to the SDKs.

**Comment by ChristianWeyer** [comment]
@yaron2 are you talking about passing raw payload to the DaprClient in the API controller of the Orders service?
If yes, then it seems this will not work?
https://github.com/dapr/dotnet-sdk/issues/773#issuecomment-953193705
